src/wallet/orchard.h

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2021 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef ZCASH_WALLET_ORCHARD_H
+#define ZCASH_WALLET_ORCHARD_H
+
+#include "rust/orchard/keys.h"
+#include "rust/orchard/wallet.h"
+#include "zcash/address/orchard.hpp"
+
+class OrchardWallet
+{
+private:
+    std::unique_ptr<OrchardWalletPtr, decltype(&orchard_wallet_free)> inner;
+
+public:
+    OrchardWallet() : inner(orchard_wallet_new(), orchard_wallet_free) {}
+    OrchardWallet(OrchardWallet&& wallet_data) : inner(std::move(wallet_data.inner)) {}
+    OrchardWallet& operator=(OrchardWallet&& wallet)
+    {
+        if (this != &wallet) {
+            inner = std::move(wallet.inner);
+        }
+        return *this;
+    }
+
+    // OrchardWallet should never be copied
+    OrchardWallet(const OrchardWallet&) = delete;
+    OrchardWallet& operator=(const OrchardWallet&) = delete;
+
+    void CheckpointNoteCommitmentTree() {
+        orchard_wallet_checkpoint(inner.get());
+    }
+
+    bool RewindToLastCheckpoint() {
+        return orchard_wallet_rewind(inner.get());
+    }
+
+    bool AddNotes(const CTransaction& tx) {
+        return orchard_wallet_add_notes_from_bundle(
+                inner.get(),
+                tx.GetHash().begin(),
+                tx.GetOrchardBundle().inner.get());
+    }
+
+    bool AppendNoteCommitments(const int nHeight, const size_t blockTxIdx, const CTransaction& tx) {
+        return orchard_wallet_append_bundle_commitments(
+                inner.get(),
+                (uint32_t) nHeight,
+                blockTxIdx,
+                tx.GetHash().begin(),
+                tx.GetOrchardBundle().inner.get()
+                );
+    }
+
+    bool IsMine(const uint256& txid) {
+        return orchard_wallet_tx_is_mine(
+                inner.get(),
+                txid.begin());
+    }
+
+    void AddSpendingKey(const libzcash::OrchardSpendingKey& sk) {
+        orchard_wallet_add_spending_key(inner.get(), sk.inner.get());
+    }
+
+    void AddFullViewingKey(const libzcash::OrchardFullViewingKey& fvk) {
+        orchard_wallet_add_full_viewing_key(inner.get(), fvk.inner.get());
+    }
+
+    void AddRawAddress(
```

Comment:
This is principally for use during wallet load, which may be out of order, so it should not fail if the IVK doesn't exist, but it can return a boolean indicating whether the IVK corresponds to a known FVK.